godot-core/src/builtin/dictionary.rs

```diff
@@ -320,6 +331,202 @@ impl Share for Dictionary {
     }
 }
 
+struct DictionaryIterator<'a> {
+    next_var: Option<Variant>,
+    dictionary: &'a Dictionary,
+}
+
+impl<'a> DictionaryIterator<'a> {
+    fn call_init(dictionary: &Dictionary) -> Option<Variant> {
+        let dictionary = dictionary.to_variant();
+        let variant: Variant = Variant::nil();
+        let mut valid = false;
+        // SAFETY:
+        // `dictionary` is a valid `Dictionary` since we have a reference to it,
+        //    so this will call the implementation for dictionaries.
+        // `variant` is an initialized and valid `Variant`.
+        // `variant_iter_init` will set `valid` to a valid `bool` value.
+        let success = unsafe {
+            interface_fn!(variant_iter_init)(
+                dictionary.var_sys(),
+                variant.var_sys(),
+                &mut valid as *mut bool as *mut u8,
+            )
+        };
+        // The function returns a bool
+        assert!(success == 0 || success == 1);
+        let success = success != 0;
+
+        if success {
+            assert!(valid);
+            Some(variant)
+        } else {
+            None
+        }
+    }
+
+    fn call_next(dictionary: &Dictionary, next_var: Variant) -> Option<Variant> {
+        let dictionary = dictionary.to_variant();
+        let mut valid = false;
+        // SAFETY:
+        // `dictionary` is a valid `Dictionary` since we have a reference to it,
+        //    so this will call the implementation for dictionaries.
+        // `next_var` is an initialized and valid `Variant`, since we own a copy of it.
+        // `variant_iter_init` will set `valid` to a valid `bool` value.
+        let success = unsafe {
+            interface_fn!(variant_iter_next)(
+                dictionary.var_sys(),
+                next_var.var_sys(),
+                &mut valid as *mut bool as *mut u8,
+            )
+        };
+        // The function returns a bool
+        assert!(success == 0 || success == 1);
+        let success = success != 0;
+
+        if success {
+            assert!(valid);
+            Some(next_var)
+        } else {
+            None
+        }
```

Comment:
This is the exact same code as above. Since `variant_iter_init` and `variant_iter_next` have the same signature, you could extract this to a function taking a `fn` function pointer.